events/iot.go

```diff
@@ -1,34 +1,46 @@
 package events
 
-// IoTCustomAuthorizerRequest contains data coming in to a custom IoT device gateway authorizer function.
-type IoTCustomAuthorizerRequest struct {
-	HTTPContext        *IoTHTTPContext `json:"httpContext,omitempty"`
-	MQTTContext        *IoTMQTTContext `json:"mqttContext,omitempty"`
-	TLSContext         *IoTTLSContext  `json:"tlsContext,omitempty"`
-	AuthorizationToken string          `json:"token"`
-	TokenSignature     string          `json:"tokenSignature"`
+// IoTCoreCustomAuthorizerRequest represents the request to an IoT Core custom authorizer.
+// See https://docs.aws.amazon.com/iot/latest/developerguide/config-custom-auth.html
+type IoTCoreCustomAuthorizerRequest struct {
+	Token              string                     `json:"token"`
+	SignatureVerified  bool                       `json:"signatureVerified"`
+	Protocols          []string                   `json:"protocols"`
+	ProtocolData       *IoTCoreProtocolData       `json:"protocolData,omitempty"`
+	ConnectionMetadata *IoTCoreConnectionMetadata `json:"connectionMetadata,omitempty"`
 }
 
-type IoTHTTPContext struct {
+type IoTCoreProtocolData struct {
+	TLS  *IoTCoreTLSContext  `json:"tls,omitempty"`
+	HTTP *IoTCoreHTTPContext `json:"http,omitempty"`
+	MQTT *IoTCoreMQTTContext `json:"mqtt,omitempty"`
+}
+
+type IoTCoreTLSContext struct {
+	ServerName string `json:"serverName"`
+}
+
+type IoTCoreHTTPContext struct {
 	Headers     map[string]string `json:"headers,omitempty"`
 	QueryString string            `json:"queryString"`
 }
 
-type IoTMQTTContext struct {
+type IoTCoreMQTTContext struct {
 	ClientID string `json:"clientId"`
 	Password []byte `json:"password"`
 	Username string `json:"username"`
 }
 
-type IoTTLSContext struct {
-	ServerName string `json:"serverName"`
+type IoTCoreConnectionMetadata struct {
+	ID string `json:"id"`
 }
 
-// IoTCustomAuthorizerResponse represents the expected format of an IoT device gateway authorization response.
-type IoTCustomAuthorizerResponse struct {
```

Comment:
I like the rename. To keep strict compile time backwards compatibility, I'd like for the old types to be left in the project too. They can have a comment like `// Deprecated does not model schema correctly, use XYZType instead`. They might also be moved to a file like `iot_deprecated.go` to help with the clutter.

Reply:
Let's not put loadbearing invariants in comments please, as they are easy to overlook.

Reply:
@bmoffatt Is it really necessary to keep the original struct? No one could be using it in practice since the `PolicyDocuments` field had the wrong type, and without that you won't have a particularly useful authorizer.

Reply:
Yeah I wanna keep the old ones, type changes and deletions are breaking changes. This isn't the first time things have been gotten wrong, and so far this project has not respond to that by deleting code. I'm not satisfied with incorrect models hanging around either, and i'd like to try and solve that holistically rather than making possibly inconsistent actions on a case-by-case basis

@LegNeato 
> Let's not put loadbearing invariants in comments please, as they are easy to overlook.

The guidance comes from https://go.dev/blog/godoc
> To signal that an identifier should not be used, add a paragraph to its doc comment that begins with “Deprecated:” followed by some information about the deprecation.

My IDE renders these with a helpful ~~strikethrough~~. The form can also be captured by the popular `staticcheck` linter

Reply:
> i'd like to try and solve that holistically rather than making possibly inconsistent actions on a case-by-case basis

Really I think this library should have made a separate module for each service. That way clients can just include the events they need rather than using a single ever-growing events package. This would also reduce some of the repetition in the typedefs (e.g., `iotcore.CustomAuthorizerResponse` vs. `events.IoTCoreCustomAuthorizerResponse`), and would allow the event structs for each service to be versioned independently.

For example in this situation we could have just made a new major rev of the IoT Core events module rather than having to (ultimately pointlessly) maintain backwards compatibility.

Reply:
Thanks for the feedback! I think it highlights some obvious shortcomings of the approach this library went with. I'll forward this on to the product team.

The approach you described is actually how it was just before the initial public release. The monolithic `events` package approached was chosen after receiving feedback from our private preview customers around poor event type discovery with IDE auto-complete suggestions at the time.